src/lib.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  Store,
+  IncrementResponse,
+  Options as RateLimitConfiguration,
+} from "express-rate-limit";
+
+import { Options, SendCommandFn } from "./types.js";
+
+/**
+ * A `Store` for the `express-rate-limit` package that stores hit counts in
+ * Redis.
+ */
+class RedisStore implements Store {
+  /**
+   * The function used to send raw commands to Redis.
+   */
+  sendCommand: SendCommandFn;
+
+  /**
+   * The text to prepend to the key in Redis.
+   */
+  prefix: string;
+
+  /**
+   * Whether to reset the expiry for a particular key whenever its hit count
+   * changes.
+   */
+  resetExpiryOnChange: boolean;
+
+  /**
+   * Stores the loaded SHA1 of the LUA script for executing the increment operations.
+   */
+  loadedScriptSha1: Promise<string>;
+
+  /**
+   * The number of milliseconds to remember that user's requests.
+   */
+  windowMs!: number;
+
+  /**
+   * @constructor for `RedisStore`.
+   *
+   * @param options {Options} - The configuration options for the store.
+   */
+  constructor(options: Options) {
+    this.sendCommand = options.sendCommand;
+    this.prefix = options.prefix ?? "rl:";
+    this.resetExpiryOnChange = options.resetExpiryOnChange ?? false;
+
+    // So that the script loading can occur non-blocking, this will send
+    // the script to be loaded, and will capture the value within the
+    // promise return. This way, if increments start being called before
+    // the script has finished loading, it will wait until it is loaded
+    // before it continues.
+    this.loadedScriptSha1 = this.loadScript();
+  }
+
+  async loadScript(): Promise<string> {
+    const result = await this.sendCommand(
+      "SCRIPT",
+      "LOAD",
+      `
+        local totalHits = redis.call("INCR", KEYS[1])
+        local timeToExpire = redis.call("PTTL", KEYS[1])
+        if timeToExpire <= 0 or ARGV[1] == "1"
+        then
+            redis.call("PEXPIRE", KEYS[1], tonumber(ARGV[2]))
+            timeToExpire = tonumber(ARGV[2])
+        end
+
+        return { totalHits, timeToExpire }
+    `
+        // Ensure that code changes that affect whitespace do not affect
+        // the script contents.
+        .replace(/^\s+/gm, "")
+        .trim()
+    );
+
+    if (typeof result !== "string") {
+      throw new TypeError("unexpected reply from redis client");
```

Comment:
I think we can rename the `passIfNotConnected` option to `passOnError`. If it is set to true, we can simply `console.log` the error, in a format similar to [this](https://github.com/nfriedly/express-rate-limit/blob/master/source/lib.ts#L110-L112) and return `{ 0, undefined }` to let the request succeed as a failover

Reply:
In this case, if an error is occurring, it's because the Redis client is not returning the correct types, and it's unlikely to continue. I would rather see the server begin to emit errors and failing when this happens rather than passing instead.

Reply:
That is a fair point - this is indeed an unexpected error and should definitely be something a developer should investigate.

However, the `passOnError` option is to ensure that the whole server doesn't come crashing down because of say, some network issue where Redis gave a blank response. I thought this was the intent behind the `passIfNotConnected` option, as mentioned in https://github.com/wyattjoh/rate-limit-redis/pull/29#issue-647529147 and https://github.com/wyattjoh/rate-limit-redis/pull/29#issuecomment-734819146.

Reply:
The `passIfNotConnected ` was explicitly used for only when the service was not connected, not that the service had an invalid response. The idea is that if it's not connected _now_, it will reconnect _soon_. I wouldn't want to enable the system to talk to an incorrectly configured Redis instance as that's something that can't be resolved. If we were to add support for the `passIfNotConnected` option, it would be for that reason, and not related to connection or operation errors.

Reply:
Good point, I guess we can pass on this option then.